apidef/oas/security.go

```diff
@@ -0,0 +1,176 @@
+		token.Signature = &Signature{}
+	}
+
+	token.Signature.Fill(authConfig)
+	if ShouldOmit(token.Signature) {
+		token.Signature = nil
+	}
+
+	if ShouldOmit(token) {
+		token = nil
+	}
+
+	s.GetTykExtension().Server.Authentication.SecuritySchemes[authConfig.Name] = token
+}
+
+func (s *OAS) extractTokenTo(api *apidef.APIDefinition, name string) {
+	var authConfig apidef.AuthConfig
+
+	if token := s.getTykTokenAuth(name); token != nil {
+		api.UseStandardAuth = token.Enabled
+		authConfig.UseCertificate = token.EnableClientCertificate
+		token.AuthSources.ExtractTo(&authConfig)
+		if token.Signature != nil {
+			token.Signature.ExtractTo(&authConfig)
+		}
+	}
+
+	s.extractSecuritySchemeTo(&authConfig)
+
+	api.AuthConfigs[apidef.AuthTokenType] = authConfig
+}
+
+func (s *OAS) extractSecurityTo(api *apidef.APIDefinition) {
+	if a := s.getTykAuthentication(); a != nil {
+		api.UseKeylessAccess = !a.Enabled
+		api.StripAuthData = a.StripAuthorizationData
+		api.BaseIdentityProvidedBy = a.BaseIdentityProvider
+	} else {
+		api.UseKeylessAccess = true
+	}
+
+	if api.AuthConfigs == nil {
+		api.AuthConfigs = make(map[string]apidef.AuthConfig)
+	}
+
+	if len(s.Security) == 0 {
+		return
+	}
+
+	for name := range s.Security[0] {
+		if s.Components.SecuritySchemes[name].Value.Type == apiKey {
+			s.extractTokenTo(api, name)
+		}
+	}
+}
+
+func (s *OAS) fillSecurity(api apidef.APIDefinition) {
+	a := s.GetTykExtension().Server.Authentication
+	if a == nil {
+		a = &Authentication{}
+		s.GetTykExtension().Server.Authentication = a
+	}
+
+	a.Enabled = !api.UseKeylessAccess
+	a.StripAuthorizationData = api.StripAuthData
+	a.BaseIdentityProvider = api.BaseIdentityProvidedBy
+
+	s.fillToken(api)
+
+	if ShouldOmit(a) {
+		s.GetTykExtension().Server.Authentication = nil
+	}
+}
+
+func (s *OAS) fillSecurityScheme(ac *apidef.AuthConfig) {
+	ss := s.Components.SecuritySchemes
+	if ss == nil {
+		ss = make(map[string]*openapi3.SecuritySchemeRef)
+		s.Components.SecuritySchemes = ss
+	}
+
+	ref, ok := ss[ac.Name]
+	if !ok {
+		ref = &openapi3.SecuritySchemeRef{
+			Value: openapi3.NewSecurityScheme(),
+		}
+		ss[ac.Name] = ref
+	}
+
+	var loc, name string
+
+	switch {
+	case ref.Value.In == header || (ref.Value.In == "" && !ac.DisableHeader):
+		loc = header
+		name = ac.AuthHeaderName
+		ac.AuthHeaderName = ""
+	case ref.Value.In == query || (ref.Value.In == "" && ac.UseParam):
+		loc = query
+		name = ac.ParamName
+		ac.ParamName = ""
+	case ref.Value.In == cookie || (ref.Value.In == "" && ac.UseCookie):
+		loc = cookie
+		name = ac.CookieName
+		ac.CookieName = ""
+	}
+
+	sec := openapi3.NewSecurityRequirement()
+	sec[ac.Name] = []string{}
+
+	found := false
+	for _, security := range s.Security {
+		if _, ok := security[ac.Name]; ok {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		s.Security.With(sec)
+	}
+
+	ref.Value.WithName(name).WithIn(loc).WithType(apiKey)
+}
+
+func (s *OAS) extractSecuritySchemeTo(ac *apidef.AuthConfig) {
```

Comment:
`extractSecuritySchemeTo` seems like a generic name, will `extractApiKeyTo` be better?

Reply:
Actually, I planned to be a generic one but it seems there will be no more cases where authconfig will be filled by security schemes.